File: 2023/day13.py
import os
import sys

test = """\
#.##..##.
..#.##.#.
##......#
##......#
..#.##.#.
..##..##.
#.#.##.#.

#...##..#
#....#..#
..##..###
#####.##.
#####.##.
..##..###
#....#..#"""

day = os.path.splitext(os.path.basename(__file__))[0]

if 'test' in sys.argv:
    data = test
else:
    data = open(day+'.txt').read()

charts = [c.splitlines() for c in data.split('\n\n')]

DEBUG = 'debug' in sys.argv

def compare_columns(chart,a,b):
    return sum(row[a] != row[b] for row in chart)

def compare_rows(chart,a,b):
    return sum(ca!=cb for ca,cb in zip(chart[a],chart[b]))

# 9 1 check 1   0 1
# 9 2 check 2   01 23
# 9 3 check 3   012 345
# 9 4 check 4   0123 4567
# 9 5 check 4   1234 5678
# 9 6 check 3   345 678 
# 9 7 check 2   56 78
# 9 8 check 1   7 8

def find_h_reflect(chart,target=0):
    W = len(chart[0])
    H = len(chart)
    for col in range(1,W):
        check = min(col, W-col)
        miss = sum( compare_columns(chart, col-1-i, col+i)
                    for i in range(check)
                )
        if miss == target:
            if DEBUG:
                print("H",col)
            return col
    return 0

def find_v_reflect(chart,target=0):
    W = len(chart[0])
    H = len(chart)
    for row in range(1,H):
        check = min(row, H-row)
        miss = sum( compare_rows(chart, row-i-1, row+i)
                    for i in range(check)     
                )
        if miss == target:
            if DEBUG:
                print("V",row)
            return row
    return 0

def part1(part,data):
    sumx = 0
    for chart in data:
        h = find_h_reflect(chart,part-1)
        v = find_v_reflect(chart,part-1)
        if DEBUG:
            print(h,v)
        sumx += 100*v+h
    return sumx

print("Part 1:", part1(1,charts))
print("Part 2:", part1(2,charts))

File: 2023/day13.go
package main

import (
	"fmt"
	"strings"

	"aoc/tools"
	_ "embed"
)

var DEBUG bool = false
var TEST bool = false
var test = `
#.##..##.
..#.##.#.
##......#
##......#
..#.##.#.
..##..##.
#.#.##.#.

#...##..#
#....#..#
..##..###
#####.##.
#####.##.
..##..###
#....#..#`[1:]

//go:embed day13.txt
var live string

func boolToInt(b bool) int {
	if b {
		return 1
	}
	return 0
}

func compare_columns(chart []string, a int, b int) int {
	sum := 0
	for _, row := range chart {
		sum += boolToInt(row[a] != row[b])
	}
	return sum
}

func compare_rows(chart []string, a int, b int) int {
	sum := 0
	for i := range len(chart[a]) {
		sum += boolToInt(chart[a][i] != chart[b][i])
	}
	return sum
}

// 9 1 check 1   0 1
// 9 2 check 2   01 23
// 9 3 check 3   012 345
// 9 4 check 4   0123 4567
// 9 5 check 4   1234 5678
// 9 6 check 3   345 678
// 9 7 check 2   56 78
// 9 8 check 1   7 8

func find_h_reflect(chart []string, target int) int {
	W := len(chart[0])
	//H := len(chart)
	for col := 1; col < W; col++ {
		miss := 0
		for i := range min(col, W-col) {
			miss += compare_columns(chart, col-1-i, col+i)
		}
		if miss == target {
			if DEBUG {
				fmt.Println("H", col)
			}
			return col
		}
	}
	return 0
}

func find_v_reflect(chart []string, target int) int {
	//W := len(chart[0])
	H := len(chart)
	for row := 1; row < H; row++ {
		miss := 0
		for i := range min(row, H-row) {
			miss += compare_rows(chart, row-1-i, row+i)
		}
		if miss == target {
			if DEBUG {
				fmt.Println("V", row)
			}
			return row
		}
	}
	return 0
}

func part1(part int, data [][]string) int {
	sum := 0
	for _, chart := range data {
		h := find_h_reflect(chart, part-1)
		v := find_v_reflect(chart, part-1)
		if DEBUG {
			fmt.Println(h, v)
		}
		sum += 100*v + h
	}
	return sum
}

func main() {
	var input string
	TEST, DEBUG, input = tools.Setup(test, live)

	// Parse the input.

	charts := [][]string{}
	for _, chunk := range strings.Split(input, "\n\n") {
		charts = append(charts, strings.Fields(chunk))
	}

	fmt.Println("Part 1:", part1(1, charts))
	fmt.Println("Part 2:", part1(2, charts))
}
